pkg/scheduler/core/scheduling_queue.go

```diff
@@ -400,35 +430,22 @@ func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
 // function adds all pods and then signals the condition variable to ensure that
 // if Pop() is waiting for an item, it receives it after all the pods are in the
 // queue and the head is the highest priority pod.
-// TODO(bsalamat): We should add a back-off mechanism here so that a high priority
-// pod which is unschedulable does not go to the head of the queue frequently. For
-// example in a cluster where a lot of pods being deleted, such a high priority
-// pod can deprive other pods from getting scheduled.
 func (p *PriorityQueue) MoveAllToActiveQueue() {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 	for _, pod := range p.unschedulableQ.pods {
-		if err := p.activeQ.Add(pod); err != nil {
-			glog.Errorf("Error adding pod %v to the scheduling queue: %v", pod.Name, err)
-		}
+		p.ScheduleWithBackoff(pod)
```

Comment:
Thanks again for this PR. It looks generally good.
I think we must start the backoff timer when a pod is added to the unschedulable queue, not when we move pods from the unschedulable to the activeQ. This would ensure that a pod that has been waiting for a long time in the unschedulable queue is not subject to further backoff when the cluster has the resources to run it.

Reply:
I thought more about this. While we still want to start the timer from the moment a pod goes to the unschedulable queue, we must ensure that if a move from the unschedulable queue to the active queue is requested, pods in the unschedulable queue go to the active queue after their backoff time expires.

One idea to address the problem is:
1. Attach a backoff timestamp to every pod that goes to the unschedulable queue.
2. Change ActiveQ to sort pods first by their "seconds to expire" and then by their priority.
3. Move all the pods, including those pods whose backoff time has not expired yet, to the ActiveQ when a move request comes in.

Reply:
Good point and I like your idea, that should simplify this as well. Thanks!

I wont be able to get to this until early next week but I'll get this fixed up then.

Reply:
I played around with this a bit and, in addition to what you say about needing to start backoff at the right time, I think more complex sorting/popping off of activeQ is needed if we want priority to effect scheduling order. If we order by backoff expiry in activeQ then this is the order we will end up scheduling if we simply activeQ.pop() to schedule (even if we have a secondary sort of priority). What we want is a priority ordered queue of everything which has completed backoff.

The naive solution for this is fairly computationally expensive - for every schedule attempt do a full activeQ scan to filter based on backoff expiry and sort this result by priority. My thinking right now is to keep a separate heap of pods ordered by backoff and keep activeQ as a pod priority ordered queue. Whenever we try to schedule we first pop all pods which have completed exipiry off our backoff queue and in to activeQ then we pop the head of activeQ (i.e. activeQ becomes a cache of priority ordered pods which have completed backoff).

Curious if you agree or if you think I am missing something?

Reply:
Sorry for the delay. I just noticed that you have replied.
I agree with you. I think your solution works well. We need to have a separate heap for pods in their backoff period. The heap is sorted by the backoff expiry. ActiveQ remains as is, i.e. sorted by priority. The algorithm is roughly like this:

1. New pods that the scheduler has never tried to schedule go directly to the ActiveQ (this logic already exists).
1. Pods that are attempted and are unschedulable are moved to the unschedulable queue and get a backoff timestamp.
1. When a move request arrives, pods whose backoff period is expired, are moved to the ActiveQ directly. The rest go to the BackoffQ.
1. A separate go routine checks periodically whether the head of the backoffQ is expired. If so, it moves it to the ActiveQ and checks the next head.